air/src/stack/stack_manipulation/mod.rs

```diff
@@ -67,12 +180,176 @@ pub fn enforce_swap_constraints<E: FieldElement>(
+
+    // Enforces the top element in the current frame is equal to the 8th element in the next
+    // frame.
+    result[5] = op_flag.movdn7() * are_equal(frame.stack_item(0), frame.stack_item_next(7));
+
+    // Enforces the top element in the current frame is equal to the 9th element in the next
+    // frame.
+    result[6] = op_flag.movdn8() * are_equal(frame.stack_item(0), frame.stack_item_next(8));
+
+    7
+}
+
+/// Enforces constraints of the CSWAP operation. The CSWAP operation swaps the elements
+/// 1,2 in the stack if the first element is 1. The stack remains the same if the top element
+/// is 0. Therefore, the following constraints are enforced:
+/// - The top element should be a binary. s0**2 - s0 = 0.
+/// - The second element in the current frame should be equal to the second element in the
+///   next frame and first element in the next frame should be equal to the third element
+///   in the current frame if the first element in the current frame is a 1.
+/// - The second element in the current frame should be equal to the first element in the
+///   next frame and second element in the next frame should be equal to the third element
+///   in the current frame if the first element in the current frame is a 0.
+pub fn enforce_cswap_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: E,
+) -> usize {
+    // Enforces the top element in the current is a binary.
+    result[0] = op_flag * is_binary(frame.stack_item(0));
+
+    // Enforces the elements 1,2 are swapped if the top element in the current frame is a 1.
+    let helper_second_0 = frame.stack_item_next(0) - frame.stack_item(0) * frame.stack_item(2);
+    let helper_first_0 = frame.stack_item(1) * binary_not(frame.stack_item(0));
+    result[1] = op_flag * are_equal(helper_second_0, helper_first_0);
+
+    // Enforces the elements 1,2 are not swapped if the top element in the current frame is a 0.
+    let helper_second_1 = frame.stack_item_next(1) - frame.stack_item(0) * frame.stack_item(1);
+    let helper_first_1 = frame.stack_item(2) * binary_not(frame.stack_item(0));
+    result[2] = op_flag * are_equal(helper_second_1, helper_first_1);
+
+    3
+}
+
+/// Enforces constraints of the CSWAPW operation. The CSWAP operation swaps the elements
+/// 1,2,3,4 with 5,6,7,8 in the stack if the first element is 1. The stack remains the same
+/// if the top element is 0. Therefore, the following constraints are enforced:
+/// - The top element should be a binary. s0**2 - s0 = 0.
+/// - If the top element is 1, stack item 1,2,3,4 have been swapped correctly.
+/// - If the top element is 0, rest of the stack should be the same.
+pub fn enforce_cswapw_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: E,
+) -> usize {
+    // Enforces the top element in the current is a binary.
+    result[0] = op_flag * is_binary(frame.stack_item(0));
```

Comment:
This is exactly the same constraint as we do for `cswap`. I wonder if there is a way to combine them somehow.

Reply:
I've combined both `cswap` and `cswapw` in one common function.